src/tiled/createtemplatetool.cpp

```diff
@@ -0,0 +1,98 @@
+/*
+ * createtemplatetool.cpp
+ * Copyright 2017, Thorbjørn Lindeijer <thorbjorn@lindeijer.nl>
+ * Copyright 2017, Mohamed Thabet <thabetx@gmail.com>
+ *
+ * This file is part of Tiled.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "createtemplatetool.h"
+
+#include "mapdocument.h"
+#include "mapobject.h"
+#include "mapobjectitem.h"
+#include "maprenderer.h"
+#include "objecttemplate.h"
+#include "snaphelper.h"
+#include "tile.h"
+#include "utils.h"
+
+using namespace Tiled;
+using namespace Tiled::Internal;
+
+CreateTemplateTool::CreateTemplateTool(QObject *parent)
+    : CreateObjectTool(parent)
+{
+    QIcon icon(QLatin1String(":images/24x24/insert-image.png"));
+    icon.addFile(QLatin1String(":images/48x48/insert-image.png"));
+    setIcon(icon);
+    Utils::setThemeIcon(this, "insert-image");
+    languageChanged();
+}
+
+void CreateTemplateTool::mouseMovedWhileCreatingObject(const QPointF &pos, Qt::KeyboardModifiers modifiers)
+{
+    const MapRenderer *renderer = mapDocument()->renderer();
+
+    const QSize imgSize = mNewMapObjectItem->mapObject()->cell().tile()->size();
```

Comment:
I have a tendency to test only tile objects and forget about the others.

I tried using the bounding rectangle dimensions to calculate the shift instead of the tile, but when clicking, each type has a different starting point, the tile objects uses the bottom left, ellipses, rectangles and text objects will use the top left, while a polyline or a polygon will use their first point.

The easy solution would be just use `pos` without any shifting. but I think we can handle them to always be centered using the bounding rectangle but with different shift for each type. Drag and drop uses the default starting point so it might be changed as well to be centered.